commands/general/stats.js

```diff
@@ -4,118 +4,165 @@ const moment = require('moment')
 
 module.exports.load = (client) => {
   client.commands['stats'] = {
-    run (message) {
-      let args = message.content.split(' ').splice(1)
-      let username
-      let discriminator
-      let mem
-      let userId
-      let av = './assets/default_avatar.jpg'
+    async run (message) {
+      let userInfo
+      try {
+        userInfo = selectTargetUser(message)
+      } catch (err) {
+        console.log(err.stack)
+        return client.errorMessage(message, `Unable to identify user: ${err}`)
+      }
+
+      try {
+        const [user, pos, tot, guild] = await Promise.all([
+          client.userRepository.load(userInfo.userId),
+          client.userRepository.getSessionPos(userInfo.userId),
+          client.userRepository.getSessionCount(),
+          client.guildRepository.load(message.guild.id)
+        ])
+        if (user !== null) {
+          const activeTime = getActiveTime(guild, userInfo.mem)
 
-      if (args.length >= 1) {
-        // fqName: "fully qualified name"
-        let fqName = args.join(' ').trim().split('#')
-        if (fqName.length === 2) {
-          username = fqName[0]
-          discriminator = fqName[1]
-          mem = client.guilds.get(message.guild.id).members.find(val => val.user.username === username && val.user.discriminator === discriminator)
-          if (mem != null) {
-            userId = mem.user.id
-            // checks if user has pfp because discord dosnt return default pfp url >:C
-            if (mem.user.avatarURL != null) {
-              av = mem.user.avatarURL
-            }
-          } else {
-            return client.errorMessage(message, `Unable to find user ${username}#${discriminator}.`)
-          }
+          userInfo.currentSession = user.current_session_playtime + activeTime
+          userInfo.overallSession = user.overall_session_playtime + activeTime
+          userInfo.rank = `${pos + 1} / ${tot}`
         } else {
-          return client.errorMessage(message, `Invalid username format.`)
-        }
-      } else {
-        username = message.author.username
-        discriminator = message.author.discriminator
-        userId = message.author.id
-        mem = client.guilds.get(message.guild.id).members.get(userId)
-        if (message.author.avatarURL != null) {
-          av = message.author.avatarURL
+          userInfo.currentSession = 0
```

Comment:
Might be worth commenting: this code is "theoretically" unreachable - we insert a user into the local repository every time a new user sends a message or drops into a VC (and I believe the welcome message is treated as a message from that user, so effectively this means we add the user as soon as they join the server). Self `p!stats` would also trigger target user creation if it hasn't happened already.

The only way this hits should be if a third party runs `p!stats` on a user that has never been heard from since the bot came up and joined while the bot was down (or before we started having a bot). In any case it might be worth shuffling the branches around so we calculate a logical time as "time obtained from repository ?? 0", regardless of whether a user object actually exists, and then return the rank as `N / A` if it's zero.